### PR TITLE
CONSOLE-2925: Add initial dynamic demo plugin tests

### PR DESCRIPTION
**Pull request by jcaianirh**
For: https://issues.redhat.com/browse/CONSOLE-2925

```
These tests are meant to:
    1. show how to test a dynamic plugin using demo as the plugin instance
    2. run locally:
      2a. build the plugin locally, and run the server
      2b. using bridge running with the plugin argument that points to the local dynanmic plugin server
      2c. will intercept the check update call to mock a toast notification that a plugin has been enabled
      2d. will not use all workload definitions defined in the yaml (not using the env variable for pull spec)
    3. run on ci:
      3a. ci will build the dynamic plugin and provide the pullspec in the env var: CYPRESS_PLUGIN_PULL_SPEC
      3b. that pull spec will be used to create the deployment on the cluster
      3c. intercepting the check-updates call will not occur as the toast will appear when the plugin is enabled
    4. the scaffolding should remain the same except modifying the constants above
```

Video of local run with bridge:
https://user-images.githubusercontent.com/35978579/145464086-5f3fba88-cbf6-4ec6-8c5e-5c7108611a1c.mp4

CI run results:
Screen Shot 2021-12-19 at 3 01 55 PM

### CODE REVIEW

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:54]: Concerned that `modal.shouldBeOpened();` will be called before the `.within click()` callback, but as long as the assertion waits for the open modal should be ok

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:70]: Why the `CHECK_UPDATE_WAIT`? I would think the wait @checkUpdates would wait for the url to be intercepted.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:155]: Can we put in a `data-test` id here?

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:162]: `data-test` id if possible

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:70]: @dtaylor113 The check-updates call happens periodically (every 15 seconds regardless), so I there needs to be a wait here since an action doesn't kick off the call as is typical with intercepting a network call with cypress.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:182]: Consider adding to `cy` commands,

```
Cypress.Commands.add('subNavShouldExist', (selector: string) =>
  cy.get('.pf-c-button.pf-m-primary.pf-m-block').should('contain', selector);
);

cy.subNavShouldExist(k8sCreate);
```

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:109]: Does this actualy work?  `isLocalDevEnvironment` is async, uses `cy.window(...).then(){...}` not sure you can rely on immediate return obj (which is a promise?).

Comment (dtaylor113) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:109]: Maybe throw isLocalDevEnviornment in an [alias](https://docs.cypress.io/guides/core-concepts/variables-and-aliases#Aliases)?

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (dtaylor113) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:144]: .nit `after[all]` is typically after `afterEach`

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts]: I restructured this code to incorporate the async command. Thanks for catching that.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts]: done

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:53]: @dtaylor113 the assertion does wait for the open model in test runs.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:98]: I would just check that cards exist instead of a specific length.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:65]: This isn't really the correct check for a local console. It's possible to run locally with auth enabled as an example.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:112]: Do we need the `isLocalDevEnvironment` check here?

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:113]: You shouldn't need `_.defaultsDeep`. Instead of `map`, I would just find the deployment and set the value.

```ts
const deployment = yamlManifest.find(({kind}) => kind === 'Deployment');
deployment.spec.template.spec.containers[0].image = PLUGIN_PULL_SPEC;
```

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:130]: I'd probably just use `oc create` to create the resources to avoid flakes and so the tests run faster. We have good coverage of the YAML editor elsewhere.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:166]: It's more important to test that the utility is working. (There should be some content and no errors.)

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (spadgett) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:181]: It's more important to test that the utilities themselves work and don't have errors.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:65]: @spadgett @dtaylor113 would this be the correct test for a local bridge env?

`if (!Cypress.config('baseUrl').includes('localhost')) `

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:112]: @spadgett we do want this check here. The plugin_pull_spec env var is only valid if we are running in ci. If we are running locally, the console-demo-plugin is a param into /bin/bridge.exe as:
`./bin/bridge -plugins console-demo-plugin=http://localhost:9001/`

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:98]: @dtaylor113 reworked these tests to look into contents instead of count and exist.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:159]: @dtaylor113 added a test-id for the consumer page title

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:65]: Added a common way to test for local bridge execution:
`export const isLocalDevEnvironment = Cypress.config('baseUrl').includes('localhost');`

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:113]: @spadgett thanks, using this to change the pull spec instead.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:130]: @spadgett ok, removed the tests for the create yaml page and used oc create instead.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:166]: @dtaylor113 added 2 data-test id's here. thanks!

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:186]: @dtaylor113 used the existing byButtonText for these thanks for the tip. Also added a test id for the title.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:166]: @spadgett tested for content and no error data

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:181]: @spadgett tested for content and no errData

**Review by andrewballantyne — APPROVED** [review]
Beautiful first set of tests... this is a big step for stability in this dynamic demo plugin!

**Review by jhadvig — APPROVED** [review]
@jcaianirh the PR looks good 👍 ... just couple of nits. No need to address. But please squash the commits :)

File: dynamic-demo-plugin/oc-manifest.yaml
Comment (jcaianirh) [reply dynamic-demo-plugin/oc-manifest.yaml:29]: This is what will be used until the release pr merges in the ci tests.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:116]: @jhadvig this was a test to see why i wasn't creating workloads from cli! You caught one leftover kubectl that i didn't change back to oc. Thanks for spotting this!

File: dynamic-demo-plugin/oc-manifest.yaml
Comment (andrewballantyne) [reply dynamic-demo-plugin/oc-manifest.yaml:29]: Honestly we should find a better home for this. QE has one too.

**Comment by jcaianirh** [comment]
QE Approver:
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

File: dynamic-demo-plugin/oc-manifest.yaml
Comment (yapei) [reply dynamic-demo-plugin/oc-manifest.yaml:29]: @jcaianirh  we have `quay.io/openshifttest/console-demo-plugin:411` image can be used, and I can also help add a `:latest` tag if needed

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (yapei) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:30]: Can we enable plugin via CLI?  that would greatly simplify the process and code..

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (yapei) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:13]: I'm hesitant adding `PLUGIN_PULL_SPEC` as cypress env since it is really specific to plugin cases while for variables becoming cypress env they should have global effect

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (yapei) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:162]: We also need remove consoleplugin/console-demo-plugin

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (yapei) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:66]: I'm not sure if hard wait is the best solution

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (yapei) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:59]: shall we intercept API request in automation from testing perspective? at least it is not normal use case IMO

Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:59]: @yapei  this is an unusual case, where we need to trigger the toast notification manually only when developing the test locally. This code is only used when creating/debugging tests locally. When running in CI we do not intercept any tests, but because the dynamic plugin is automatically loaded when running ./bin/bridge locally with the plugins param this intercept is needed to mock the ci environment.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:66]: Hard wait is needed because there is poll interval that we need to accommodate during testing. The toast is not triggered when the change occurs, rather it is triggered after polling the system twice and comparing the plugin values. Thats why a wait is needed in this case @yapei

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:162]: @yapei yes thats correct. adding / removing the plugin is on line 76 of the enable/disable plugin function that is called on line 160.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:13]: @yapei yes we need this env variable because there will be another pr in the release repository. That pr is going to set the image pull spec in the environment variable for this test. The environment variable will be set during the builds for e2e testing in that release repository. Thats why we are grabbing that pull spec here. You will see that pr coming soon and see how the pull spec is passed through the env var.

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:30]: @yapei we are creating our workloads via the oc create cli commands. We want to test the gui and the ability to set up these plugins via the gui and not the cli. I think here we want to make sure we are testing the gui.

**Comment by yapei** [comment]
looks good to me

**Comment by jcaianirh** [comment]
Thanks @yapei could you add the qe-approved label please ?

File: frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts
Comment (jhadvig) [inline frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:13]: Should this be `CYPRESS_PLUGIN_PULL_SPEC` ?

Comment (jcaianirh) [reply frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.spec.ts:13]: So the cypress.env api is as follows: https://docs.cypress.io/guides/guides/environment-variables#Option-3-CYPRESS_

I think we are ok with this definition.

**Pull request merged**
